src/Hamiltonians/HOCartesian.jl

```diff
@@ -0,0 +1,417 @@
+    pairs::SVector{N,T}
+    length::Int
+end
+
+function OccupiedPairsMap(addr::SingleComponentFockAddress{N}) where {N}
+    omm = OccupiedModeMap(addr)
+    T = eltype(omm)
+    P = N * (N - 1) ÷ 2
+    pairs = MVector{P,Tuple{T,T}}(undef)
+    L = 0
+    for i in eachindex(omm)
+        p_i = omm[i]
+        if p_i.occnum > 1
+            L += 1
+            @inbounds pairs[L] = (p_i, p_i)
+        end
+        for j in 1:i-1
+            p_j = omm[j]
+            L += 1
+            @inbounds pairs[L] = (p_i, p_j)
+        end
+    end
+    
+    return OccupiedPairsMap(SVector(pairs), L)
+end
+
+Base.size(opm::OccupiedPairsMap) = (opm.length,)
+function Base.getindex(opm::OccupiedPairsMap, i)
+    @boundscheck 1 ≤ i ≤ opm.length || throw(BoundsError(opm, i))
+    return opm.pairs[i]
+end
+
+"""
+    HOCartesian(addr; S, η, g = 1.0, interaction_only = false)
+
+Implements a harmonic oscillator in Cartesian basis with contact interactions.
+```math
+\\hat{H} = \\sum_{i} \\epsilon_i n_i + \\frac{g}{2}\\sum_{ijkl} V_{ijkl} a^†_i a^†_j a_k a_l
+```
+Indices ``i, \\ldots`` are ``D``-tuples for a ``D``-dimensional harmonic oscillator. 
+The energy scale is defined by the first dimension i.e. ``\\hbar \\omega_x`` so that 
+single particle energies are ``\\epsilon_i = (i_x + 1/2) + \\eta_y (i_y+1/2) + \\ldots``.
+The factors ``\\eta_y, \\ldots`` allow for anisotropic trapping geometries and are assumed to 
+be greater than `1` so that ``\\omega_x`` is the smallest trapping frequency.
+Matrix elements ``V_{ijkl}`` are for a contact interaction calculated in this basis using 
+first-order perturbation theory. All states with the same noninteracting energy are connected
+by this interaction.
+
+# Arguments
+
+* `addr`: the starting address, defines number of particles and total number of modes.
+* `S`: Tuple of the number of levels in each dimension, including the groundstate. The 
+    allowed couplings between states is defined by the aspect ratio of `S .- 1`. Defaults 
+    to a 1D spectrum with number of levels matching modes of `addr`. Will be sorted to make 
+    the first dimension the largest.
+* `η`: Define a custom aspect ratio for the trapping potential strengths, instead of deriving
+    from `S .- 1`. This will only affect the single particle energy scale and not the 
+    interactions. The values are always scaled relative to the first dimension, which sets 
+    the energy scale of the system, ``\\hbar\\omega_x``.
+* `g`: the (isotropic) interparticle interaction parameter. The value of `g` is assumed 
+    to be in trap units.
+* `interaction_only`: if set to `true` then the noninteracting single-particle terms are 
+    ignored. Useful if only energy shifts due to interactions are required.
+
+See also [`HOCartesianSeparable`](@ref).
+
+# Warning: 
+`num_offdiagonals` is a bad estimate for this Hamiltonian. Take care when building 
+a matrix (use option `col_hint` with [`BasisSetRep`](@ref)) or using QMC methods.
```

Comment:
```suggestion
!!! warning
    `num_offdiagonals` is a bad estimate for this Hamiltonian. Take care when building 
    a matrix (use option `col_hint` with [`BasisSetRep`](@ref)) or using QMC methods.
```
This will show the warning in a yellow box in the docs.